src/server/pegasus_hotkey_collector.cpp

```diff
@@ -0,0 +1,175 @@
+    _collector_state.store(COARSE);
+    return true;
+}
+
+void hotkey_collector::clear()
+{
+    _coarse_data_collector.reset();
+    _fine_data_collector.reset();
+    _collector_state.store(STOP);
+}
+
+std::string hotkey_collector::get_status()
+{
+    collector_state_set status = _collector_state.load();
+    if (status == STOP)
+        return "STOP";
+    if (status == COARSE)
+        return "COARSE";
+    if (status == FINE)
+        return "FINE";
+    return "FINISH";
+}
+
+bool hotkey_collector::get_result(std::string &result)
+{
+    if (_collector_state.load() != FINISH)
+        return false;
+    result = _fine_result;
+    return true;
+}
+
+void hotkey_collector::capture_msg_data(dsn::message_ex **requests, const int count)
+{
+    if (_collector_state.load() == STOP || count == 0) {
+        return;
+    }
+    for (int i = 0; i < count; i++) {
+        ::dsn::blob key;
+        if (requests[i] != nullptr) {
+            dsn::task_code rpc_code(requests[0]->rpc_code());
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_MULTI_PUT) {
+                dsn::apps::multi_put_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_INCR) {
+                dsn::apps::incr_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.key;
+                capture_blob_data(key);
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_CHECK_AND_SET) {
+                dsn::apps::check_and_set_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_CHECK_AND_MUTATE) {
+                dsn::apps::check_and_mutate_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_PUT) {
+                dsn::apps::update_request thrift_request;
+                unmarshall(requests[i], thrift_request);
+                requests[i]->restore_read();
+                key = thrift_request.key;
+                capture_blob_data(key);
+            }
+        }
+    }
+}
+
+void hotkey_collector::capture_blob_data(const ::dsn::blob &key)
+{
+    if (_collector_state.load() == STOP) {
+        return;
+    }
+    std::string hash_key, sort_key;
+    if (key.length() < 2)
```

Comment:
You can assume that every key's length > 2 here.

Reply:
My unit test made some empty strings, should I delete them?

Reply:
blob type key size should >= 2, even if it's empty string.